datafusion/common/src/utils/memory.rs

```diff
@@ -0,0 +1,123 @@
+// under the License.
+
+//! This module provides a function to estimate the memory size of a HashTable prior to alloaction
+
+/// Estimates the memory size required for a hash table prior to allocation.
+///
+/// # Parameters
+/// - `num_elements`: The number of elements expected in the hash table.
+/// - `fixed_size`: A fixed overhead size associated with the collection
+/// (e.g., HashSet or HashTable).
+/// - `T`: The type of elements stored in the hash table.
+///
+/// # Details
+/// This function calculates the estimated memory size by considering:
+/// - An overestimation of buckets to keep approximately 1/8 of them empty.
+/// - The total memory size is computed as:
+///   - The size of each entry (`T`) multiplied by the estimated number of
+///     buckets.
+///   - One byte overhead for each bucket.
+///   - The fixed size overhead of the collection.
+/// - Returns `usize::MAX` if an overflow occurs.
+///
+/// # Examples
+/// ---
+///
+/// ## From within a struct
+///
+/// ```rust
+/// use datafusion_common::utils::memory::estimate_memory_size;
+///
+/// struct MyStruct<T> {
+///     values: Vec<T>,
+///     other_data: usize,
+/// }
+///
+/// impl<T> MyStruct<T> {
+///     fn size(&self) -> usize {
+///         let num_elements = self.values.len();
+///         let fixed_size = std::mem::size_of_val(self) +
+///           std::mem::size_of_val(&self.values);
+///
+///         estimate_memory_size::<T>(num_elements, fixed_size)
+///     }
+/// }
+/// ```
+/// ---
+/// ## With a simple collection
+///
+/// ```rust
+/// use datafusion_common::utils::memory::estimate_memory_size;
+/// use std::collections::HashMap;
+///
+/// let num_rows = 100;
+/// let fixed_size = std::mem::size_of::<HashMap<u64, u64>>();
+/// let estimated_hashtable_size =
+///   estimate_memory_size::<(u64, u64)>(num_rows,fixed_size);
+/// ```
+pub fn estimate_memory_size<T>(num_elements: usize, fixed_size: usize) -> usize {
+    // For the majority of cases hashbrown overestimates the bucket quantity
+    // to keep ~1/8 of them empty. We take this factor into account by
+    // multiplying the number of elements with a fixed ratio of 8/7 (~1.14).
+    // This formula leads to overallocation for small tables (< 8 elements)
+    // but should be fine overall.
+    num_elements
+        .checked_mul(8)
+        .and_then(|estimate| estimate.checked_div(7))
+        .map(|buckets| buckets.next_power_of_two())
+        .and_then(|estimated_buckets| {
+            // + size of entry * number of buckets
+            // + 1 byte for each bucket
+            // + fixed size of collection (HashSet/HashTable)
+            std::mem::size_of::<T>()
+                .checked_mul(estimated_buckets)
+                .and_then(|size_of_entries| {
+                    estimated_buckets
+                        .checked_add(fixed_size)
+                        .and_then(|overhead| size_of_entries.checked_add(overhead))
+                })
+        })
```

Comment:
I found this formulation very hard to follow due to all the unesting

I wonder if we could use the `?` operator to make it read better


Something like

```rust
    num_elements
        .checked_mul(8)
        .and_then(|estimate| estimate.checked_div(7))
        .map(|buckets| buckets.next_power_of_two())
        .and_then(|estimated_buckets| {
            // + size of entry * number of buckets
            std::mem::size_of::<T>()
                .checked_mul(estimated_buckets)?
                // + 1 byte for each bucket
                .checked_add(estimated_buckets)?
                // + fixed size of collection (HashSet/HashTable)
                .checked_add(fixed_size)
        })
        .unwrap_or(usize::MAX)
```

Reply:
Wouldn't this also require us to change the fn to Result<usize>, which leads to a change in the Accumulator trait.

Reply:
I don't think so (I actually tried that code above and it works fine)

In rust `?` can also be used in functions that return `Option`  (to return `None`)

Reply:
thanks for the suggestion; this looks way nicer 🚀